kratos/tests/cpp_tests/utilities/test_parallel_utilities.cpp

```diff
@@ -0,0 +1,210 @@
+    double reference_sum = 0.0;
+    double reference_sub = 0.0;
+    for(auto item : data_vector)
+    {
+        reference_max = std::max(reference_max, item);
+        reference_min = std::min(reference_min, item);
+        reference_sum += item;
+        reference_sub -= item;
+    }
+    class CustomReducer{
+        public:
+            typedef std::tuple<double,double> value_type;
+            double max_value = -std::numeric_limits<double>::max();
+            double max_abs = 0.0;
+
+            value_type GetValue()
+            {
+                value_type values;
+                std::get<0>(values) = max_value;
+                std::get<1>(values) = max_abs;
+                return values;
+            }
+
+            void LocalReduce(double function_return_value){
+                this->max_value = std::max(this->max_value,function_return_value);
+                this->max_abs   = std::max(this->max_abs,std::abs(function_return_value));
+            }
+            void ThreadSafeReduce(CustomReducer& rOther){
+                #pragma omp critical
+                {
+                this->max_value = std::max(this->max_value,rOther.max_value);
+                this->max_abs   = std::max(this->max_abs,std::abs(rOther.max_abs));
+                }
+            }
+    };
+
+    auto partition = IndexPartition<unsigned int>(data_vector.size());
+
+    double max_value,max_abs;
+    std::tie(max_value,max_abs) = partition.for_reduce<CustomReducer>([&](unsigned int i){
+            return data_vector[i]; //note that here the lambda returns the values to be reduced
+        });
+
+    KRATOS_CHECK_EQUAL(max_value, 0.0 );
+    KRATOS_CHECK_EQUAL(max_abs, nsize-1 );
+
+
+
+    //******************************************************************************************
+    //here we reduce at once, the sum,min,max and sub
+    typedef CombinedReduction< SumReduction<double>,
+                               MinReduction<double>,
+                               MaxReduction<double>,
+                               SubReduction<double>
+            > MultipleReduction;
+
+    //auto reduction_res
+    double sum,min,max,sub;
+    std::tie(sum,min,max,sub) = IndexPartition<unsigned int>(data_vector.size()).
+        for_reduce<MultipleReduction>(
+            [&](unsigned int i){
+                    double to_sum = data_vector[i];
+                    double to_max = data_vector[i];
+                    double to_min = data_vector[i];
+                    double to_sub = data_vector[i];
+                    return std::make_tuple( to_sum, to_max, to_min, to_sub ); //note that these may have different types
+                }
+            );
+    KRATOS_CHECK_EQUAL(sum, reference_sum );
+    KRATOS_CHECK_EQUAL(min, reference_min );
+    KRATOS_CHECK_EQUAL(max, reference_max );
+    KRATOS_CHECK_EQUAL(sub, reference_sub );
+}
+
+KRATOS_TEST_CASE_IN_SUITE(OmpVsPureC11, KratosCoreFastSuite)
```

Comment:
should this remain a test?

Reply:
i would leave it for now, mostly as it marks how it should be done once in the next step. Anyway doing a replace of this function would be very easy